### PR TITLE
Unify parameterize api

### LINKED ISSUE #982: Parametrize API inconsistency
Our default strategy for parametrized tests is fuzzing. I believe there may be a benefit in making it explicit.

Argument pro this change is it increases the discoverability of the framework.
If you always write `given`, it gives users not familiar with protostar a keyword they can look for in documentation, and read about fuzzing. Without it, they are left wondering what is happening, and where those parameters come from.
It makes an interface a little more consistent as well.

On the other hand it may sometimes create extra boilerplate.

Core question is, how often is fuzzing used without any strategy?

@mkaput @karol-bisztyga

### PR DESCRIPTION
**Pull request by karol-bisztyga**
When it comes to `given` and `example` cheatcodes, we want to consider the following scenarios:
- no `given`, no `examples` - for now, it is going to be marked as deprecated, but in the future, an error should be raised
- only `given` - just use data from `given`
- only `example`s - use only `example`s
- `given` and `example`s - use data from `given` plus `example`s

fix #982

### CODE REVIEW

File: protostar/testing/environments/fuzz_test_execution_environment.py [hunk 1]
```diff
@@ -59,6 +61,17 @@ async def execute(self, function_name: str) -> FuzzTestExecutionResult:
         database = InMemoryExampleDatabase()
         runs_counter = RunsCounter(budget=self.state.config.fuzz_max_examples)
 
+        if (
+            not self.state.config.fuzz_examples
+            and not self.state.config.fuzz_declared_strategies
+        ):
+            self._logger.warning(
+                "not providing the test data is deprecated and will break test cases in the future releases, "
```
Comment (MaksymilianDemitraszek) [inline protostar/testing/environments/fuzz_test_execution_environment.py:69]:
```suggestion
                "Not providing the test parameters is deprecated and will break test cases in the future releases, "
```

File: protostar/testing/environments/fuzz_test_execution_environment.py [hunk 2]
```diff
@@ -59,6 +61,17 @@ async def execute(self, function_name: str) -> FuzzTestExecutionResult:
...
+                "please use one of the following cheatcodes in the setup function in order to "
+                "explicitly provide data to be tested: \n- example\n- given"
+            )
+            # todo raise an error in future releases
```
Comment (MaksymilianDemitraszek) [inline protostar/testing/environments/fuzz_test_execution_environment.py:73]: Can we create task for that on github instead of this TODO?

Comment (MaksymilianDemitraszek) [reply protostar/testing/environments/fuzz_test_execution_environment.py:73]: comments tend to be forgotten :p

File: tests/integration/fuzzing/fuzzing_test.py [hunk 3]
```diff
@@ -134,16 +135,21 @@ async def test_parameterized_with_examples_tests(
     assert_cairo_test_cases(
         testing_summary,
         expected_passed_test_cases_names=[
-            "test_examples",
-            "test_examples_without_given",
             "test_check_exact_example",
+            "test_given_and_examples",
+            "test_only_examples",
+            "test_only_given",
+            "test_no_data",  # should be broken in the future by raising an error
```
Comment (MaksymilianDemitraszek) [inline tests/integration/fuzzing/fuzzing_test.py:142]:
```suggestion
            "test_no_data",
```

File: tests/integration/fuzzing/fuzzing_test.py [hunk 4]
```diff
@@ -134,16 +135,21 @@ async def test_parameterized_with_examples_tests(
...
+    assert len(testing_summary.passed) == 5
     passed_set = {
         getattr(passed, "fuzz_runs_count")
         for passed in testing_summary.passed
         if hasattr(passed, "fuzz_runs_count")
     }
-    assert passed_set == {1, 2, 7}
+    # TestConfig().fuzz_max_examples is a default value for max examples
+    # it is going to be applied for test_only_given
+    # because max_examples is not provided there, so it falls back to the default value
```
Comment (MaksymilianDemitraszek) [inline tests/integration/fuzzing/fuzzing_test.py:154]:
```suggestion
    # TestConfig().fuzz_max_examples is a default value for max examples
```

File: tests/integration/fuzzing/parameterized_test.cairo [hunk 5]
```diff
@@ -2,43 +2,79 @@
 
 from starkware.cairo.common.math import assert_le
 
+////////// no data - broken test
```
Comment (MaksymilianDemitraszek) [inline tests/integration/fuzzing/parameterized_test.cairo:5]:
```suggestion
```

File: tests/integration/fuzzing/parameterized_test.cairo [hunk 6]
```diff
@@ -2,43 +2,79 @@
...
+func setup_no_data() {
+    return ();
+}
+
+@external
+func test_no_data{syscall_ptr: felt*, range_check_ptr}(a: felt, b: felt) {
+    return ();
+}
 
+////////// only examples
```
Comment (MaksymilianDemitraszek) [inline tests/integration/fuzzing/parameterized_test.cairo:17]:
```suggestion
```
It is redundant with function name

File: tests/integration/fuzzing/parameterized_test.cairo [hunk 7]
```diff
@@ -2,43 +2,79 @@
...
+
+@external
+func test_only_given{syscall_ptr: felt*, range_check_ptr}(a: felt, b: felt) {
+    assert_le(0, a);
+    assert_le(0, b);
+    assert_le(a, b);
+    return ();
+}
+
+////////// given and examples
```
Comment (MaksymilianDemitraszek) [inline tests/integration/fuzzing/parameterized_test.cairo:54]:
```suggestion
```
It is redundant with function name

File: tests/integration/fuzzing/parameterized_test.cairo [hunk 8]
```diff
@@ -2,43 +2,79 @@
...
 @external
-func test_examples_without_given{syscall_ptr: felt*, range_check_ptr}(a: felt, b: felt) {
+func test_given_and_examples{syscall_ptr: felt*, range_check_ptr}(a: felt, b: felt) {
     assert_le(0, a);
     assert_le(0, b);
     assert_le(a, b);
     return ();
 }
 
+////////// check exact example
```
Comment (MaksymilianDemitraszek) [inline tests/integration/fuzzing/parameterized_test.cairo:76]:
```suggestion
```

File: website/docs/tutorials/07-testing/02-cheatcodes/example.md [hunk 9]
```diff
@@ -18,6 +18,9 @@ This cheatcode is only available in [setup cases](../README.md#setup-case).
 :::warning
 `example` is not limited by [`max_examples`](./max-examples.md) and is not connected to it in any way.
 :::
+:::warning
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/02-cheatcodes/example.md:21]: We have to update fuzzing tutorial, default fuzzing is mentioned there in many places

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Good job, most comments are nitpicks but we have to update all fuzzing tutorials

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md [hunk 10]
```diff
@@ -35,3 +35,6 @@ func test_less_equal_compare{syscall_ptr : felt*, range_check_ptr}(a : felt, b :
 :::info
 There is also the [`example`](./example.md) cheatcode that tests explicitly provided cases. It may come handy if you don't want to rely on randomization.
 :::
+:::warning
+You have to provide the testing data explicitly either by using `example` or [`given`](./given.md).
```
Comment (mkaput) [inline website/docs/tutorials/07-testing/02-cheatcodes/given.md:39]:
```suggestion
You have to provide the testing data explicitly either by using [`example`](./example.md) or `given`.
```

File: tests/integration/fuzzing/parameterized_test.cairo (hunk 5 above)
Comment (karol-bisztyga) [reply tests/integration/fuzzing/parameterized_test.cairo]: [modify cairo test style](https://github.com/software-mansion/protostar/pull/995/commits/fe79f3f1bc58c4a01034e9b71e3b04621944ba90)

File: tests/integration/fuzzing/parameterized_test.cairo [hunk 11]
```diff
@@ -2,43 +2,79 @@
 
 from starkware.cairo.common.math import assert_le
 
+////////// no data - broken test
+
 @external
-func setup_examples() {
-    %{
-        max_examples(5)
+func setup_no_data() {
```
[the remark this answers is not in the document]
Comment (karol-bisztyga) [reply tests/integration/fuzzing/parameterized_test.cairo]: [modify cairo test style](https://github.com/software-mansion/protostar/pull/995/commits/fe79f3f1bc58c4a01034e9b71e3b04621944ba90)

File: tests/integration/fuzzing/parameterized_test.cairo (hunk 6 above)
Comment (karol-bisztyga) [reply tests/integration/fuzzing/parameterized_test.cairo]: [modify cairo test style](https://github.com/software-mansion/protostar/pull/995/commits/fe79f3f1bc58c4a01034e9b71e3b04621944ba90)

File: tests/integration/fuzzing/parameterized_test.cairo (hunk 7 above)
Comment (karol-bisztyga) [reply tests/integration/fuzzing/parameterized_test.cairo]: [modify cairo test style](https://github.com/software-mansion/protostar/pull/995/commits/fe79f3f1bc58c4a01034e9b71e3b04621944ba90)

File: tests/integration/fuzzing/parameterized_test.cairo (hunk 8 above)
Comment (karol-bisztyga) [reply tests/integration/fuzzing/parameterized_test.cairo]: [modify cairo test style](https://github.com/software-mansion/protostar/pull/995/commits/fe79f3f1bc58c4a01034e9b71e3b04621944ba90)

File: protostar/testing/environments/fuzz_test_execution_environment.py (hunk 2 above)
Comment (karol-bisztyga) [reply protostar/testing/environments/fuzz_test_execution_environment.py:73]: #1008

Comment (karol-bisztyga) [reply protostar/testing/environments/fuzz_test_execution_environment.py]: [remove todo comment](https://github.com/software-mansion/protostar/pull/995/commits/a087126c773de3184e464240861927027df7c323)

File: website/docs/tutorials/07-testing/02-cheatcodes/example.md (hunk 9 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/02-cheatcodes/example.md:21]: [update the fuzzing tutorial](https://github.com/software-mansion/protostar/pull/995/commits/cf20cf1d265a8ff397e6487a07546b7b2946fab2) I tried to cover this but might've missed something, let me know, thanks!

File: protostar/testing/environments/fuzz_test_execution_environment.py [hunk 12]
```diff
@@ -59,6 +61,16 @@ async def execute(self, function_name: str) -> FuzzTestExecutionResult:
...
         runs_counter = RunsCounter(budget=self.state.config.fuzz_max_examples)
 
+        if (
+            not self.state.config.fuzz_examples
+            and not self.state.config.fuzz_declared_strategies
+        ):
+            self._logger.warning(
+                "Not providing the test parameters is deprecated and will break test cases in the future releases, "
+                "please use one of the following cheatcodes in the setup function in order to "
+                "explicitly provide data to be tested: \n- example\n- given"
```
Comment (MaksymilianDemitraszek) [inline protostar/testing/environments/fuzz_test_execution_environment.py:71]:
```suggestion
                "explicitly provide test data: \n- example\n- given"
```

File: protostar/testing/environments/fuzz_test_execution_environment.py [hunk 13]
```diff
@@ -59,6 +61,16 @@ async def execute(self, function_name: str) -> FuzzTestExecutionResult:
         database = InMemoryExampleDatabase()
         runs_counter = RunsCounter(budget=self.state.config.fuzz_max_examples)
 
+        if (
+            not self.state.config.fuzz_examples
+            and not self.state.config.fuzz_declared_strategies
+        ):
+            self._logger.warning(
+                "Not providing the test parameters is deprecated and will break test cases in the future releases, "
+                "please use one of the following cheatcodes in the setup function in order to "
```
Comment (MaksymilianDemitraszek) [inline protostar/testing/environments/fuzz_test_execution_environment.py:70]: Nit
```suggestion
                "Please use one of the following cheatcodes in the case setup function in order to "
```

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md [hunk 14]
```diff
@@ -28,8 +28,10 @@ func test_integers{syscall_ptr : felt*, range_check_ptr}(a : felt, b : felt) {
 ```
 
 :::note
-By default, Protostar applies the [`strategy.felts()`](#strategyfelts) strategy to all felt
-parameters.
+In the past, Protostar used to apply the [`strategy.felts()`](#strategyfelts) strategy to all felt
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/03-fuzzing/strategies.md:31]: I think we shouldn't mention the past in the docs. We should just describe it as it is. It makes it easier to read.

File: website/docs/tutorials/07-testing/02-cheatcodes/example.md [hunk 15]
```diff
@@ -18,6 +18,9 @@ This cheatcode is only available in [setup cases](../README.md#setup-case).
...
+You have to provide the testing data explicitly either by using `example` or [`given`](./given.md).
+:::
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/02-cheatcodes/example.md:23]: This is a weird information in current context.
```suggestion

```

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md [hunk 16]
```diff
@@ -35,3 +35,6 @@ func test_less_equal_compare{syscall_ptr : felt*, range_check_ptr}(a : felt, b :
 :::info
 There is also the [`example`](./example.md) cheatcode that tests explicitly provided cases. It may come handy if you don't want to rely on randomization.
 :::
+:::warning
+You have to provide the testing data explicitly either by using [`example`](./example.md) or `given`.
+:::
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/02-cheatcodes/given.md:40]:
```suggestion
```

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
There are still leftovers in fuzzing docs, like:

> Protostar tests can take parameters, which makes such tests to be run in a fuzzing mode. In this mode, Protostar treats the test case parameters as a specification of the test case...

I think you need to read these docs carefuly and restucture. Remember about updating cairo examples as well.

File: website/docs/tutorials/07-testing/02-cheatcodes/example.md (hunk 15 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/02-cheatcodes/example.md]: [Address direct comments](https://github.com/software-mansion/protostar/pull/995/commits/d53af654b1789f34245a4e363740e3035ff1cfb5)

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md (hunk 14 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md]: [Address direct comments](https://github.com/software-mansion/protostar/pull/995/commits/d53af654b1789f34245a4e363740e3035ff1cfb5)

File: protostar/testing/environments/fuzz_test_execution_environment.py (hunk 12 above)
Comment (karol-bisztyga) [reply protostar/testing/environments/fuzz_test_execution_environment.py]: [Address direct comments](https://github.com/software-mansion/protostar/pull/995/commits/d53af654b1789f34245a4e363740e3035ff1cfb5)

File: website/docs/tutorials/07-testing/03-fuzzing/README.md [hunk 17]
```diff
@@ -237,7 +237,7 @@ Protostar offers various strategies tailored for specific use cases, check out
 the [Strategies](./strategies.md) page for more information.
 Associating a fuzzing strategy to a parameter is done using the [`given`](../02-cheatcodes/given.md)
 cheatcode, which is only available within [setup cases][setup-case].
-The default strategy [`felts`](./strategies.md#strategyfelts) accepts a keyword argument `rc_bound`
+The strategy [`felts`](./strategies.md#strategyfelts) accepts a keyword argument `rc_bound`
```
Comment (mkaput) [inline website/docs/tutorials/07-testing/03-fuzzing/README.md:240]:
```suggestion
The [`felts`](./strategies.md#strategyfelts) strategy accepts a keyword argument `rc_bound`
```

I think this is more English?

File: website/docs/tutorials/07-testing/03-fuzzing/README.md [hunk 18]
```diff
@@ -287,10 +287,6 @@ Each resource counter presents a summary of observed values across all test runs
 
 ## Adjusting fuzzing quality
 
-By default, Protostar tries to fail a test case within 100 examples.
-The default value is chosen to suit a workflow where the test will be part of a suite that is
-regularly executed locally or on a CI server,
-balancing total running time against the chance of missing a bug.
```
Comment (mkaput) [inline website/docs/tutorials/07-testing/03-fuzzing/README.md:293]: Why is this removed? It's the only place in docs which says what's the default...

**Review by mkaput — CHANGES_REQUESTED** [review]

File: website/docs/tutorials/07-testing/03-fuzzing/README.md (hunk 17 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/README.md:260]: right

File: website/docs/tutorials/07-testing/03-fuzzing/README.md (hunk 18 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/README.md:293]: right, sorry

Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/README.md:293]: [revert adjusting fuzzing quality](https://github.com/software-mansion/protostar/pull/995/commits/1d12293ba0b243f635e23ddff1d9552d3dc25aa2)

**Review by mkaput — APPROVED** [review]
LGTM

**Comment by karol-bisztyga** [comment]
> LGTM

Thx, will wait for @MaksymilianDemitraszek then

File: website/docs/tutorials/07-testing/03-fuzzing/README.md [hunk 19]
```diff
@@ -116,6 +123,22 @@ Protostar will run any test that takes parameters in fuzz testing mode, so let's
 test:
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/03-fuzzing/README.md:123]: This sentence is not reworked properly

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md [hunk 20]
```diff
@@ -28,8 +28,8 @@ func test_integers{syscall_ptr : felt*, range_check_ptr}(a : felt, b : felt) {
 ```
 
 :::note
-By default, Protostar applies the [`strategy.felts()`](#strategyfelts) strategy to all felt
-parameters.
+It is required to either provide a strategy using [`given`](../02-cheatcodes/given.md)
+or specific values for parameters using [`example`](../02-cheatcodes/example.md).
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/03-fuzzing/strategies.md:32]: This is not a properly delivered information. If `given` is necessary then we shouldn't put it in a note but into main section. Please remove this note.

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Fuzzing and strategies are still written in a confisuing way with some leftovers from the old version. 
e.g.
> Fuzzer allows to provide such assumptions to avoid testing against values which are not covered by the code. 

Right now fuzzer always requires them, so this sentence is misleading.

And some other different quirks. I suggest just rewriting those sections from scratch, otherwise you will not avoid sentences which are weird and misleading in the context of current implementation.

File: website/docs/tutorials/07-testing/03-fuzzing/README.md (hunk 19 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/README.md:123]: Sorry but I don't get which sentence, can you specify? You commented on an unchanged line preceded by unchanged text.

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md (hunk 20 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md:32]: sure

File: website/docs/tutorials/07-testing/03-fuzzing/README.md (hunk 19 above)
Comment (MaksymilianDemitraszek) [reply website/docs/tutorials/07-testing/03-fuzzing/README.md:123]: >Protostar will run any test that takes parameters in fuzz testing mode, so let's rewrite our unit
test:

Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/README.md:123]: I see, I just misunderstood, what you meant was "not reworked at all but it should've been"

**Comment by karol-bisztyga** [comment]
> Fuzzing and strategies are still written in a confisuing way with some leftovers from the old version. e.g.
> 
> > Fuzzer allows to provide such assumptions to avoid testing against values which are not covered by the code.
> 
> Right now fuzzer always requires them, so this sentence is misleading.

This was discussed offline and we decided to leave it as is as it is correct. The "assumption" does not refer to using `example`/`given`, it is about `assert`s.

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md (hunk 20 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md:31]: [docs](https://github.com/software-mansion/protostar/pull/995/commits/ece63bb97c024245dd553f1c0720d0305c96b92c)

File: website/docs/tutorials/07-testing/03-fuzzing/README.md (hunk 19 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/README.md]: [docs updates](https://github.com/software-mansion/protostar/pull/995/commits/4521eb3d29990ea5d54633e16da1a8aa36b9c511)

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md [hunk 21]
```diff
@@ -27,10 +27,8 @@ func test_integers{syscall_ptr : felt*, range_check_ptr}(a : felt, b : felt) {
 }
 ```
 
-:::note
-By default, Protostar applies the [`strategy.felts()`](#strategyfelts) strategy to all felt
-parameters.
-:::
+It is required to either provide a strategy using [`given`](../02-cheatcodes/given.md)
+or specific values for parameters using [`example`](../02-cheatcodes/example.md).
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/03-fuzzing/strategies.md:31]: I wonder if this sentence is necessary

**Review by MaksymilianDemitraszek — APPROVED** [review]

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md (hunk 21 above)
Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md:31]: We can remove it

Comment (karol-bisztyga) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md]: [remove sentence](https://github.com/software-mansion/protostar/pull/995/commits/4179ef53bc46bfb439929a359de8cc2c3c7a3d76)
